crates/bevy_pbr/src/render/prepass.rs

```diff
@@ -0,0 +1,599 @@
+                targets,
+            })
+        } else {
+            None
+        };
+
+        // Use the vertex shader from the material if present
+        let vert_shader_handle = if let Some(handle) = &self.material_vertex_shader {
+            handle.clone()
+        } else {
+            PREPASS_SHADER_HANDLE.typed::<Shader>()
+        };
+
+        let mut descriptor = RenderPipelineDescriptor {
+            vertex: VertexState {
+                shader: vert_shader_handle,
+                entry_point: "vertex".into(),
+                shader_defs,
+                buffers: vec![vertex_buffer_layout],
+            },
+            fragment,
+            layout: Some(bind_group_layout),
+            primitive: PrimitiveState {
+                topology: key.mesh_key.primitive_topology(),
+                strip_index_format: None,
+                front_face: FrontFace::Ccw,
+                cull_mode: None,
+                unclipped_depth: false,
+                polygon_mode: PolygonMode::Fill,
+                conservative: false,
+            },
+            depth_stencil: Some(DepthStencilState {
+                format: DEPTH_PREPASS_FORMAT,
+                depth_write_enabled: true,
+                depth_compare: CompareFunction::GreaterEqual,
+                stencil: StencilState {
+                    front: StencilFaceState::IGNORE,
+                    back: StencilFaceState::IGNORE,
+                    read_mask: 0,
+                    write_mask: 0,
+                },
+                bias: DepthBiasState {
+                    constant: 0,
+                    slope_scale: 0.0,
+                    clamp: 0.0,
+                },
+            }),
+            multisample: MultisampleState {
+                count: key.mesh_key.msaa_samples(),
+                mask: !0,
+                alpha_to_coverage_enabled: false,
+            },
+            label: Some("prepass_pipeline".into()),
+        };
+
+        // This is a bit risky because it's possible to change something that would
+        // break the prepass but be fine in the main pass.
+        // Since this api is pretty low-level it doesn't matter that much, but it is a potential issue.
+        M::specialize(&self.material_pipeline, &mut descriptor, layout, key)?;
+
+        Ok(descriptor)
+    }
+}
+
+// Extract the render phases for the prepass
+pub fn extract_camera_prepass_phase(
+    mut commands: Commands,
+    cameras_3d: Extract<
+        Query<
+            (
+                Entity,
+                &Camera,
+                Option<&DepthPrepass>,
+                Option<&NormalPrepass>,
+            ),
+            With<Camera3d>,
+        >,
+    >,
+) {
+    for (entity, camera, depth_prepass, normal_prepass) in cameras_3d.iter() {
+        if camera.is_active {
+            let mut entity = commands.get_or_spawn(entity);
+            entity.insert((
+                RenderPhase::<Opaque3dPrepass>::default(),
```

Comment:
Should this be added only if one of the prepass components is Some?

Reply:
Yeah, good call